src/Crypto/Lol/Cyclotomic/Tensor.hs

```diff
@@ -54,37 +54,37 @@ class (TElt t Double, TElt t (Complex Double))
   type TElt t r :: Constraint
 
   -- | Properties that hold for any index. Use with '\\'.
-  entailIndexT :: Tagged (t m r)
-                  (Fact m :- (Applicative (t m), Traversable (t m)))
+  entailIndexT :: Tagged (t m r) (Fact m :- (Applicative (t m), Traversable (t m)))
   
   -- | Properties that hold for any (legal) fully-applied tensor. Use
   -- with '\\'.
-  entailFullT :: Tagged (t m r)
-                 ((Fact m, TElt t r) :- 
-                  (Eq (t m r), ZeroTestable (t m r), Ring (t m r), 
-                   NFData (t m r), Random (t m r)))
+  entailEqT :: Tagged (t m r) ((Fact m, TElt t r, Eq r) :- (Eq (t m r)))
+  entailZTT :: Tagged (t m r) ((Fact m, TElt t r, ZeroTestable r) :- (ZeroTestable (t m r)))
+  entailRingT :: Tagged (t m r) ((Fact m, TElt t r, Ring r) :- (Ring (t m r)))
+  entailNFDataT :: Tagged (t m r) ((Fact m, TElt t r, NFData r) :- (NFData (t m r)))
+  entailRandomT :: Tagged (t m r) ((Fact m, TElt t r, Random r) :- (Random (t m r)))
 
   -- | Converts a scalar to a tensor in the powerful basis
-  scalarPow :: (Fact m, TElt t r) => r -> t m r
+  scalarPow :: (Fact m, TElt t r, Additive r) => r -> t m r
 
   -- | 'l' converts from decoding-basis representation to
   -- powerful-basis representation; 'lInv' is its inverse.
-  l, lInv :: (Fact m, TElt t r) => t m r -> t m r
+  l, lInv :: (Fact m, TElt t r, Additive r) => t m r -> t m r
 
   -- | Multiply by @g@ in the powerful/decoding basis
-  mulGPow, mulGDec :: (Fact m, TElt t r) => t m r -> t m r
+  mulGPow, mulGDec :: (Fact m, TElt t r, Additive r) => t m r -> t m r
 
   -- | Divide by @g@ in the powerful/decoding basis.  The 'Maybe'
   -- output indicates that the operation may fail, which happens
   -- exactly when the input is not divisible by @g@.
-  divGPow, divGDec :: (Fact m, TElt t r) => t m r -> Maybe (t m r)
+  divGPow, divGDec :: (Fact m, TElt t r, ZeroTestable r, IntegralDomain r) => t m r -> Maybe (t m r)
```

Comment:
There are a lot of `ZT r, ID r` constraints here and below; seems very duplicitous.  What can be done?